### PR TITLE
Generate the bci-base container via dockerfile generator as well

### PR DESCRIPTION
**Pull request by dirkmueller**

### CODE REVIEW

**Comment by dirkmueller** [comment]
(this needs a bit more work)

File: src/bci_build/package/basecontainers.py
Comment (dcermak) [inline src/bci_build/package/basecontainers.py:40]: The dedent does nothing
```suggestion
    return r"""
```

File: src/bci_build/package/basecontainers.py
Comment (dcermak) [inline src/bci_build/package/basecontainers.py:98]: This could be replaced with our `LOG_CLEAN` constant

File: src/bci_build/package/__init__.py
Comment (dcermak) [inline src/bci_build/package/__init__.py:938]: The `replaces=` attribute has been deleted in schemaversion 6.7:
```
    Add XSL stylesheet to convert from v6.6 to v6.7
    
    Delete the following obsolete and unhandled attributes and sections
    
    * <image ... kiwirevision="..."/>
    * <type ... checkprebuilt="..."/>
    * <type ... fsnocheck="..."/>
    * <repository ... status="..."/>
    * <repository ... prefer-license="..."/>
    * <package replaces="..."/>
    * <defaultprebuilt>
    * <defaultdestination>
    * <defaultroot>
    * <partitioner>
    * <rpm-force>
```

File: src/bci_build/package/basecontainers.py
Comment (dirkmueller) [reply src/bci_build/package/basecontainers.py]: I would like to first produce a 100% identical result with the handcrafted config.sh. we can clean this up then in a followup step.

File: src/bci_build/package/__init__.py
Comment (dirkmueller) [reply src/bci_build/package/__init__.py]: This has been fixed now.

File: src/bci_build/package/basecontainers.py
Comment (dirkmueller) [reply src/bci_build/package/basecontainers.py]: replaced with jinja.template rendering.

**Comment by Vogtinator** [comment]
Looking at the diff, some issues:

- [ ] Duplicate `suseImportBuildKey`
- [ ] Duplicate `/var/log` cleaning. The manual one has better comments.
- [ ] Duplicate `Remove locale files`
- [ ] `<label name="com.suse.image-type" value="sle-bci"/>` got lost

**Comment by dirkmueller** [comment]
> Looking at the diff, some issues:
> * [ ]  `<label name="com.suse.image-type" value="sle-bci"/>` got lost

This is intentional, we have deprecated that label and removed it from all BCI containers.

**Comment by Vogtinator** [comment]
In the case that we might get rid of findutils in the container again, we'd have to switch back to the globstar based locale deletion

**Comment by dirkmueller** [comment]
>  Duplicate /var/log cleaning. The manual one has better comments.

I can add the better comments to the sources in dockerfile generator?

> Duplicate Remove locale files

removed.

**Comment by dirkmueller** [comment]
> In the case that we might get rid of findutils in the container again, we'd have to switch back to the globstar based locale deletion

removed find(1) usage again.

File: src/bci_build/package/__init__.py
Comment (dcermak) [inline src/bci_build/package/__init__.py:38]: You cannot do this, `LOG_CLEAN` must be a single line instruction, because it is used in all `Dockerfile` templates inside of `RUN` instructions.

Comment (dirkmueller) [reply src/bci_build/package/__init__.py]: restored

File: src/bci_build/package/base/README.md.j2
Comment (dcermak) [inline src/bci_build/package/base/README.md.j2:29]: Leftover?
```suggestion
```

File: src/bci_build/templates.py
Comment (dcermak) [inline src/bci_build/templates.py:95]: Do we need this? I don't see it in [sles15-image.kiwi](https://build.suse.de/projects/SUSE:SLE-15-SP6:Update/packages/sles15-image/files/sles15-image.kiwi?expand=1) but there's this instead:
```xml
<!-- OBS-ExcludeArch: i586 -->
```

Comment (Vogtinator) [reply src/bci_build/templates.py:95]: Previous kiwi schemaversions allow `<package ... replaces="foobar"/>` but that's no longer part of the schema, so it has to be done a different way

Comment (Vogtinator) [reply src/bci_build/templates.py:95]: The `<!-- OBS-ExcludeArch: i586 -->` is only needed if this is part of the SLE project itself which provides the i586 arch as well.

File: src/bci_build/package/base/README.md.j2
Comment (alexandrevicenzi) [inline src/bci_build/package/base/README.md.j2:11]: nit: you could remove _many_ from _across many different environments_

File: src/bci_build/package/base/README.md.j2
Comment (alexandrevicenzi) [inline src/bci_build/package/base/README.md.j2:12]: which templates and tools? it is not clear here.

File: src/bci_build/package/base/README.md.j2
Comment (alexandrevicenzi) [inline src/bci_build/package/base/README.md.j2:13]: what can be used? BCI? the templates and tools from the previous sentence? which is also not clear.

File: src/bci_build/package/base/README.md.j2
Comment (alexandrevicenzi) [inline src/bci_build/package/base/README.md.j2:17]: nit: it would be nice to mention what security and compliance are being supported here, as we know what we support.

File: src/bci_build/package/base/README.md.j2
Comment (alexandrevicenzi) [inline src/bci_build/package/base/README.md.j2:24]: nit: it would be nice to make it clear that Linux OS here is the host OS of the container runtime, otherwise it gives an impression that you can run BCI with another OS inside.

File: src/bci_build/templates.py
Comment (dcermak) [reply src/bci_build/templates.py:95]: > Previous kiwi schemaversions allow `<package ... replaces="foobar"/>` but that's no longer part of the schema, so it has to be done a different way

I see, thanks!

> The `<!-- OBS-ExcludeArch: i586 -->` is only needed if this is part of the SLE project itself which provides the i586 arch as well.

Don't we want to publish the container from the SLE project though? Then we'd need this additional comment

File: src/bci_build/package/base/README.md.j2
Comment (dirkmueller) [reply src/bci_build/package/base/README.md.j2:17]: I'd like to address this in a followup. this one is intended to as closely as possible resemble the existing sle15-image. we can clean it up once this PR has been merged. it is large enough as is.

File: src/bci_build/package/base/README.md.j2
Comment (dirkmueller) [reply src/bci_build/package/base/README.md.j2:24]: I'd like to address this in a followup. this one is intended to as closely as possible resemble the existing sle15-image. we can clean it up once this PR has been merged. it is large enough as is.

File: src/bci_build/package/base/README.md.j2
Comment (dirkmueller) [reply src/bci_build/package/base/README.md.j2:12]: I'd like to address this in a followup. this one is intended to as closely as possible resemble the existing sle15-image. we can clean it up once this PR has been merged. it is large enough as is.

File: src/bci_build/package/base/README.md.j2
Comment (dirkmueller) [reply src/bci_build/package/base/README.md.j2:11]: I'd like to address this in a followup. this one is intended to as closely as possible resemble the existing sle15-image. we can clean it up once this PR has been merged. it is large enough as is.

Comment (alexandrevicenzi) [reply src/bci_build/package/base/README.md.j2:11]: ok, fair.

**Review by alexandrevicenzi — DISMISSED** [review]

**Comment by Vogtinator** [comment]
I'm looking at the resulting diff of sles15-image and I found the following:

```diff
 #======================================
-# Exclude docs intallation
+# Exclude docs installation
 #--------------------------------------
-sed -i 's/.*rpm.install.excludedocs.*/# rpm.install.excludedocs = yes/g' /etc/zypp/zypp.conf
+sed -i 's/.*rpm.install.excludedocs.*/rpm.install.excludedocs = yes/g' /etc/zypp/zypp.conf
```

i.e. previously excludedocs was not actually set to yes, ouch. IMO ok to fix, but might need discussion and documentation. This hits derived containers only.

**Comment by Vogtinator** [comment]
```diff
-    <version>15.0.0</version>
+    <version>15.5.0</version>
```

will reset the cicount, meaning tags would start from 15.5.1.0 again.

**Comment by Vogtinator** [comment]
```
+<!-- OBS-AddTag: suse/sle15:%OS_VERSION_ID_SP% suse/sle15:%OS_VERSION_ID_SP%.%RELEASE% bci/bci-base:%OS_VERSION_ID_SP% bci/bci-base:%OS_VERSION_ID_SP%.%RELEASE% -->
```

but also

```
+      <containerconfig name="suse/sle15" tag="%OS_VERSION_ID_SP%" maintainer="SUSE LLC (https://www.suse.com/)" additionaltags="%OS_VERSION_ID_SP%,%OS_VERSION_ID_SP%.%RELEASE%">
```

i.e. the `suse/sle15` tags are set twice.

**Comment by Vogtinator** [comment]
`<package name="patterns-base-minimal_base"/>` is new in bootstrap packages. Intentional?

**Comment by dirkmueller** [comment]
> i.e. the `suse/sle15` tags are set twice.

yes, luckily OBS de-duplicates them, so it's not a real issue.

**Comment by dirkmueller** [comment]
> `<package name="patterns-base-minimal_base"/>` is new in bootstrap packages. Intentional?

not new, there since

```
------------------------------------------------------------------- 
Mon Oct 23 13:34:14 UTC 2023 - Dirk Müller <dmueller@suse.com>      
                                                                    
- add patterns-base-minimal_base to have the SLE branding        
```

**Comment by Vogtinator** [comment]
> > i.e. the `suse/sle15` tags are set twice.
> 
> yes, luckily OBS de-duplicates them, so it's not a real issue.

Can we avoid the duplication? Could be as simple as removing `additionaltags` from the template. FWICT there's a bug anyway:

```xml
      <containerconfig
          name="{{ image.build_tags[0].split(':')[0] }}"
          tag="{{ image.build_tags[0].split(':')[1] }}"
          maintainer="{{ image.maintainer }}"{% if image.kiwi_additional_tags %}
{%- if image.entrypoint_user  %}
          user="{{ image.entrypoint_user }}"
{%- endif %}
          additionaltags="{{ image.kiwi_additional_tags }}"{% endif %}>
        <labels>
```

If `image.kiwi_additional_tags` is empty, the user is not set either.

**Comment by dirkmueller** [comment]
> ```diff
> -    <version>15.0.0</version>
> +    <version>15.5.0</version>
> ```
> 
> will reset the cicount, meaning tags would start from 15.5.1.0 again.

yes, we can set that via buildrelease though. for now I've restricted it to sp6 and set a buildrelease matching that.

**Comment by Vogtinator** [comment]
> > `<package name="patterns-base-minimal_base"/>` is new in bootstrap packages. Intentional?
> 
> not new, there since
> 
> ```
> ------------------------------------------------------------------- 
> Mon Oct 23 13:34:14 UTC 2023 - Dirk Müller <dmueller@suse.com>      
>                                                                     
> - add patterns-base-minimal_base to have the SLE branding        
> ```

Not there in SUSE:SLE-15-SP5:Update/sles15-image yet

**Comment by dirkmueller** [comment]
> If `image.kiwi_additional_tags` is empty, the user is not set either.

good catch, moved to https://github.com/SUSE/BCI-dockerfile-generator/pull/1419 

in regards to reducing the duplication, I'll take a look later.

File: src/bci_build/templates.py
Comment (dcermak) [inline src/bci_build/templates.py:95]: Can we please not have hacks like this in the template? Image specific settings belong into the image and not into the general purpose template imho

File: src/bci_build/package/__init__.py
Comment (dcermak) [inline src/bci_build/package/__init__.py:208]: Rebase artifact

**Comment by dirkmueller** [comment]
> > > `<package name="patterns-base-minimal_base"/>` is new in bootstrap packages. Intentional?
> Not there in SUSE:SLE-15-SP5:Update/sles15-image yet

correct, but we'll only do the SP6 image for now. but I've added a condition for it it on SP6

File: src/bci_build/templates.py
Comment (dirkmueller) [reply src/bci_build/templates.py]: added a separate kiwi_ignore_packages flag for this.

File: src/bci_build/package/__init__.py
Comment (dirkmueller) [reply src/bci_build/package/__init__.py]: yeah.. removed.

**Comment by dirkmueller** [comment]
> Can we avoid the duplication? Could be as simple as removing `additionaltags` from the template.

I'd rather go the other way and remove OBS specific hacks so that builds outside obs work better. moved that to #1423 (which is not working atm because of some bug somewhere)

for now, this is a cosmetical  problem, I would ignore it in order to not block this change from going in further.

**Comment by Vogtinator** [comment]
There's duplication even in the kiwi tag:

```
+          tag="%OS_VERSION_ID_SP%"
+          additionaltags="%OS_VERSION_ID_SP%,%OS_VERSION_ID_SP%.%RELEASE%"
```

i.e. `suse/sle15:%OS_VERSION_ID_SP%` is defined three times now :-/

**Comment by Vogtinator** [comment]
Looks good otherwise.

> I'd rather go the other way and remove OBS specific hacks so that builds outside obs work better. moved that to https://github.com/SUSE/BCI-dockerfile-generator/pull/1423 (which is not working atm because of some bug somewhere)

Where: OBS, kiwi, dockerfile-generator?

File: src/bci_build/package/base/README.md.j2
Comment (Vogtinator) [reply src/bci_build/package/base/README.md.j2:29]: Looks like it's back?

Having an empty section looks really ugly

**Comment by dirkmueller** [comment]
> There's duplication even in the kiwi tag:

> i.e. `suse/sle15:%OS_VERSION_ID_SP%` is defined three times now :-/

Yes, but it doesn't matter. Yet, I fixed it now.

**Review by alexandrevicenzi — APPROVED** [review]

**Review by dcermak — APPROVED** [review]
LGTM, but I'd like @Vogtinator to give his ACK here

**Comment by dirkmueller** [comment]
> > I'd rather go the other way and remove OBS specific hacks so that builds outside obs work better. moved that to #1423 (which is not working atm because of some bug somewhere)
> Where: OBS, kiwi, dockerfile-generator?

I believe a bug in kiwi, but I haven't checked it closely (could be elsewhere). I can see the skopeo call to add the additional names, but they are still not there on the build result. maybe also OBS interaction. needs to be debugged tehre.

**Review by Vogtinator — APPROVED** [review]
Looking at the diff between SUSE:SLE-15-SP6:Update and this I don't see anything obvious wrong anymore, just mostly cosmetic issues.

**Pull request merged**
